Automattic/wp-calypso
# composite-checkout: Allow adding temporary products to the cart

#### Changes proposed in this Pull Request

This PR refactors the `useShoppingCart` hook so that when its `addItem` function is used to add an item to the cart, it is added as a fully fleshed-out, but with null data, response cart object. This way, while the cart is being synchronized with the server, all the rendering machinery will still operate with the optimistically-added new cart item. 

Prior to this change, if a cart item was added with the typical data available for adding new cart items (generally just `product_id` and `product_slug`), the page would throw an error as it tried to convert the missing data into objects suitable for the UI. With this change, a placeholder item (with the label `…`) will be displayed instead. In a future PR we can improve the UI for these placeholder items.

#### Testing instructions

This PR should not change anything, and is extracted from #41456. There is currently no code that uses the `addItem` function. If you'd like you can merge this branch with the commits in #41456 and perform the testing instructions there to see these changes in action.

Load composite checkout and verify that there are no issues. Remove an item from the cart and verify that it gets removed successfully. Add items to the cart via URL (eg: by clicking the purchase buttons on the plans page) and verify that the item gets added successfully.
